src/Build/Logging/BinaryLogger/Postprocessing/ArchiveFile.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.IO.Compression;
+using System.Text;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// An object model for binlog embedded files.
+    /// </summary>
+    public sealed class ArchiveFile
+    {
+        private bool _streamAcquired;
+        private bool _stringAcquired;
+        private readonly StreamReader _contentReader;
+        private string? _content;
+
+        public ArchiveFile(string fullPath, Stream contentStream)
+        {
+            FullPath = fullPath;
+            _contentReader = new StreamReader(contentStream);
+        }
+
+        public ArchiveFile(string fullPath, string content)
```

Comment:
This is for cases where api user wants to edit the embedded file (same use case as with string events) - in such case they'll set the new instance (instead of altering it in place - as I didn't want this to be mutable).

Btw. the OM model was a bit simplified from the last time you reviewed (partly inspired by your proposal to split the different behavior in subclasses)